### ISSUE TITLE
Randomizer softlocks at the 4 pokemon encounter in front of Kincaid's lab during the night event at Ranger School

### ISSUE THREAD

**Issue by pokemonlover1234** [ronitsinha/almia-randomizer#1]
When doing the night event during the Ranger School, I made it to the basement to drop off the 4 stylers. I encountered a Gardevoir, Dusknoir, Mismagius, and Banette. After the red exclamation point appears above Keith's head, the game softlocks and does not enter the encounter.

**Comment by pokemonlover1234** [comment]
Image for reference
![19624788-7d36-4e5b-8445-b5b347090ff4](https://github.com/ronitsinha/almia-randomizer/assets/18287167/a15c25bb-d912-4c67-86d0-440e3ebe021f)

**Comment by ronitsinha** [comment]
Thanks for bringing this to my attention. I'll take a look at this when I get the chance. When you say the game softlocks, do you mean the game just freezes, or does something else happen?

**Comment by pokemonlover1234** [comment]
The music continues to play, the Pokemon's sprites continue their idle animations and the map cursor continues to pulse but otherwise the game is stuck. No button input does anything.

**Comment by pokemonlover1234** [comment]
As long as I'm here, I had a couple questions and I'm unsure if I should open another issue for them. Whenever something is fixed, I assume I'll need to rerandomize the ROM. Do you know if saves from prior randomizations will be compatible or would I need to start over? Also, does the randomizer code guarantee all Pokemon in the glossary will appear somewhere in the world (i.e. does it guarantee 100% is possible)?

**Comment by ronitsinha** [comment]
From my testing, re-randomizing will preserve save-states (like on an emulator), I'll do some more testing to see how it works with in-game saves. As for the 100%, right now it is not guaranteed by the randomizer. This is because the randomizer will replace a pokemon with a pokemon of the same field move, with the same power or higher (so none of the puzzles are impossible). In order to guarantee 100%, I'd most likely have to restrict randomization to some extent.

**Comment by pokemonlover1234** [comment]
If you were willing to implement that in some way, it could be an option on the executable, but it's up to you.

**Comment by ronitsinha** [comment]
It's something I'm considering, I would definitely love to have more user-customization options in the tool.

**Comment by pokemonlover1234** [comment]
For future reference, if I have questions/suggestions, should I just make a new issue?

**Comment by ronitsinha** [comment]
Yeah that's fine. I am a bit busy these days so it might take me some time to get to each issue, but yeah feel free to make new ones.

**Comment by ronitsinha** [comment]
Just a quick update on this; There are a few places in the game where it seems like randomizing the pokemon will cause the game to freeze. One such example is the catching tutorial. I'm gonna try testing this by modifying the rom with another tool (AlmiaE) and seeing if the game still freezes. If so, then there might just be some parts of the game that cannot be randomized without significantly more work.

**Comment by pokemonlover1234** [comment]
My guess is that cutscene-triggered encounters (such as the tutorial+basement ones that cause the softlock) are somehow looking for a specific hard-coded pokemon to trigger the encounter. 

I'm planning on trying to partially reverse engineer the game to see why exactly the softlock happens (from the disassembler in Desmume, it seems to be caught in an infinite loop) and see if there is some way to modify the ROM to fix it. 

That will probably take months though. If AlmiaE works and this is unnecessary, great! If not, the set of rooms with custscene encounters should be temporarily excluded from the randomizer (as you did with the catch tutorial). If I do figure out a way to fix it through the reverse engineering effort, I'll probably need a list of the problemaric encounters anyway.

**Comment by DarkTails02** [comment]
Hello i have the same issue as pokemonlover1234 the game is stuck while the encounter of the basement what should i do (i encounter 4 ghost pokemon but it do the same thing)

**Comment by ronitsinha** [comment]
So I've done some more digging. I haven't experimented with the ghost pokemon in the basement, but I've instead been looking at the target clear tutorial with Ms. Claire, since the Bidoof in that cutscene couldn't be randomized without errors either (if you look at the code, you'll see `m001_013.map.dat` is excluded for this reason). Essentially, the way this program currently works is by randomizing the pokemon id's stored in the `.map.dat` files, but it turns out that there are scripts (`.fsb` files) that need to be modified as well. Leaving them unmodified results in the game crashing/freezing like we see here.

In the example of the catching tutorial: one of the map scripts (`m001.fsb`) tries to get a reference to a Pokemon map object based on a Pokemon id. In this case it's Bidoof, with id 0xAD (see [this](https://github.com/SunakazeKun/AlmiaE/blob/master/docs/list_mons.txt) for all poke ids). When we randomized the Bidoof tho, the script cannot find the map object with Bidoof's poke id, which is what causes erratic behavior.

I think we can solve this if, during the randomization, we look not only at the maps but also the corresponding scripts to make corresponding changes. I'm no longer on break tho, so implementing this might be slow-going. Feel free to ask me more questions about this and I can give further detail.

Also, for more info on how FSB scripts work, I found this incredible repo: https://gitlab.com/Fexean/almiascript

**Comment by DetectiveAlice** [comment]
2025, having the same issue outside of kincaid's door.

**Comment by Jerome1998** [comment]
Ran into the same issue today :(
